onmt/model_builder.py

```diff
@@ -157,62 +150,52 @@ def build_base_model(model_opt, fields, gpu, checkpoint=None):
         the NMTModel.
     """
     assert model_opt.model_type in ["text", "img", "audio"], \
-        ("Unsupported model type %s" % (model_opt.model_type))
+        "Unsupported model type %s" % model_opt.model_type
 
     # for backward compatibility
     if model_opt.rnn_size != -1:
         model_opt.enc_rnn_size = model_opt.rnn_size
         model_opt.dec_rnn_size = model_opt.rnn_size
-        if model_opt.model_type == 'text' and \
-           model_opt.enc_rnn_size != model_opt.dec_rnn_size:
-                raise AssertionError("""We do not support different encoder and
-                                     decoder rnn sizes for translation now.""")
 
     # Build encoder.
     if model_opt.model_type == "text":
-        src_dict = fields["src"].vocab
-        feature_dicts = inputters.collect_feature_vocabs(fields, 'src')
-        src_embeddings = build_embeddings(model_opt, src_dict, feature_dicts)
-        encoder = build_encoder(model_opt, src_embeddings)
+        feat_fields = [fields[k]
+                       for k in inputters.collect_features(fields, 'src')]
+        src_emb = build_embeddings(model_opt, fields["src"], feat_fields)
+        encoder = build_encoder(model_opt, src_emb)
     elif model_opt.model_type == "img":
-        if ("image_channel_size" not in model_opt.__dict__):
+        # why is the model_opt.__dict__ check necessary?
+        if "image_channel_size" not in model_opt.__dict__:
             image_channel_size = 3
         else:
             image_channel_size = model_opt.image_channel_size
 
-        encoder = ImageEncoder(model_opt.enc_layers,
-                               model_opt.brnn,
-                               model_opt.enc_rnn_size,
-                               model_opt.dropout,
-                               image_channel_size)
+        encoder = ImageEncoder(
+            model_opt.enc_layers, model_opt.brnn, model_opt.enc_rnn_size,
+            model_opt.dropout, image_channel_size)
     elif model_opt.model_type == "audio":
-        encoder = AudioEncoder(model_opt.rnn_type,
-                               model_opt.enc_layers,
-                               model_opt.dec_layers,
-                               model_opt.brnn,
-                               model_opt.enc_rnn_size,
-                               model_opt.dec_rnn_size,
-                               model_opt.audio_enc_pooling,
-                               model_opt.dropout,
-                               model_opt.sample_rate,
-                               model_opt.window_size)
+        encoder = AudioEncoder(
```

Comment:
vertically aligned parameters are easier to read, I would keep it.

Reply:
On this point I somewhat disagree. Arguments lists longer than four or five elements are intrinsically difficult to read no matter how you align them. I'm open to vertical, but if so I think it's better to have the arguments start on the line below the function name because 1) there's less gaping whitespace on the left and 2) you don't have to adjust the whole column of arguments if the name of the function changes length.

Reply:
In this case at least we instantly recognize that `model_opt.` is repeated among each parameters.

No problem on starting on next line. The only thing is, when defining functions, parameters would need to be indented twice to avoid `E125 continuation line with same indent as next logical line`, which isn't very common.

Reply:
I think it's perfectly clear that all the arguments come from `model_opt` regardless, but I'm not sure how relevant this is to readability and I'm not convinced it's a useful insight. I can't look at a function with ten positional arguments and tell if I've transposed two of them no matter how they're spaced. Doing it horizontally in multiple lines allows one to at least group related arguments by putting them on the same line, which is impossible if it's just a single column.